docs/developer-guide/ksqldb-rest-api/streaming-endpoint.md

```diff
@@ -0,0 +1,147 @@
+---
+layout: page
+title: HTTP Streaming API
+tagline: streaming endpoints
+description: The HTTP Streaming API lets you execute pull or push queries and stream inserts to the
+server
+keywords: ksqlDB, query, insert, select
+---
+
+Please note: These endpoints are used by the ksqlDB Java client. If you are using Java you might want
+to consider using the Java client rather than using this API directly.
+
+These endpoints are only available when using HTTP 2.
+
+### Executing pull or push queries
+
+The request method will be a POST.
+
+Requests will be made to the endpoint "/query-stream".
+
+The body of the request is a JSON object UTF-8 encoded as text, containing the arguments for the
+operation (newlines have been added here for the sake of clarity but the real JSON must not contain
+ unescaped newlines)
+
+````
+{
+"sql": "select * from foo", <----- the SQL of the query to execute
+"properties": {             <----- Optional properties for the query
+    "prop1": "val1",
+    "prop2": "val2"
+   }
+}
+
+````
+
+The endpoint produces responses with two possible content types `application/json` and
+`application/vnd.ksqlapi.delimited.v1`. To choose one or the other you should set the `Accept`
+header in the request. The default is `application/vnd.ksqlapi.delimited.v1`.
+
+In the case of a successful query, if the content type is `application/vnd.ksqlapi.delimited.v1`
+then the results are returned as a header JSON object followed by zero or more JSON arrays
+delimited by newlines. Newline delimited formats are easy to parse by clients, and don't require
+a streaming JSON parser on the client in the case that intermediate results need to be output.
+
+````
+{
+"queryId", "xyz123",                          <---- unique ID of the query, used when terminating the query
+"columnNames":["col", "col2", "col3"],        <---- the names of the columns
+"columnTypes":["BIGINT", "STRING", "BOOLEAN"] <---- The types of the columns
+}
+````
+
+Followed by zero or more JSON arrays:

```

Comment:
Can we also add a note about the error response format? Specifically, the rows are JSON arrays unless an error is encountered while streaming, in which case a JSON object of a particular format is returned. I know last time we said this doesn't appear to actually happen even though the infrastructure is set up for it, but it'd still be useful to call out since clients should handle this case.

Reply:
I don't think this can happen so doesn't make sense to me to add it to the docs.

Reply:
Will we want to support errors in this way in the future? If not we should remove the functionality from the server to clean up the code. If so we should have a plan for how to update clients to support these errors if the error mechanism is added in the future. (The Java client already supports them, but authors of any new clients won't if the mechanism isn't documented.) It seems easier to have clients support the mechanism up front rather than having to go through and add support in the clients when we want to start using the mechanism.